awx/main/models/schedules.py

```diff
@@ -35,6 +35,45 @@
 UTC_TIMEZONES = {x: tzutc() for x in dateutil.parser.parserinfo().UTCZONE}
 
 
+SECONDS_IN_WEEK = 7 * 24 * 60 * 60
+
+
+def fast_forward_date(rrule):
+    '''
+    Utility to fast forward an rrule, maintaining consistency in the resulting
+    occurrences
+    Fast forwards the rrule to 7 days ago
+    Returns a datetime object
+    '''
+    if not rrule._freq in (dateutil.rrule.HOURLY, dateutil.rrule.MINUTELY):
+        raise RuntimeError("Cannot fast forward rrule, frequency must be HOURLY or MINUTELY")
+
+    interval = rrule._interval if rrule._interval else 1
+    if rrule._freq == dateutil.rrule.HOURLY:
+        interval *= 60 * 60
+    elif rrule._freq == dateutil.rrule.MINUTELY:
+        interval *= 60
+
+    if type(interval) == float and not interval.is_integer():
```

Comment:
Is it intentional that `isinstance(interval, float)` is not used here? Do we want to disallow subclasses?
Also, wouldn't it be more correct to do an identity check, as in `type(interval) is float`?

Looking at what's checked, I think that it should be possible to use `rrule._interval` instead and move it to the beginning of the function, having a chance to bail before the computations. This is probably not a huge performance benefit, but it would improve the structure by grouping all the “guard expression”-style checks together.

Additionally, once these checks are close, they could be moved into a helper called smth like `_assert_rrule_is_fast-forwardable()`, which might improve the structure even further.

Reply:
made some recent code changes around this